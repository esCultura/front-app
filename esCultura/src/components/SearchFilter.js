import React, {useState, useEffect} from 'react';
import { StyleSheet, View, TextInput, Text, Modal, Pressable, Button} from 'react-native';
import { CheckBox, Slider} from 'react-native-elements';
import DateTimePicker from '@react-native-community/datetimepicker';
import FilterLeft from 'react-native-bootstrap-icons/icons/filter-left';
import CalendarEvent from 'react-native-bootstrap-icons/icons/calendar-event';
import ArrowDown from 'react-native-bootstrap-icons/icons/arrow-down';

export default function SearchFilter({onVariableChange}, isList) {

    const [eventsData, setEventsData] = useState([]);
    const [endPoint, setEndPoint] = useState('');

    const [modalVisible, setModalVisible] = useState(false);
    const [slider1, setSlider1] = useState(0);

    const [dateIni, setDateIni] = useState(new Date());
    const [dateFi, setDateFi] = useState(new Date());
    const [showDatePickerIni, setShowDatePickerIni] = useState(false);
    const [showDatePickerFi, setShowDatePickerFi] = useState(false);

    const [checkTeatre, setCheckTeatre] = useState(false);
    const [checkEspectacles, setCheckEspectacles] = useState(false);
    const [checkMusica, setCheckMusica] = useState(false);
    const [checkExposicions, setCheckExposicions] = useState(false);
    const [checkCinema, setCheckCinema] = useState(false);
    const [checkDivulgacio, setCheckDivulgacio] = useState(false);
    const [checkDansa, setCheckDansa] = useState(false);
    const [checkInfantil, setCheckInfantil] = useState(false);
    

    const [textSearch, setTextSearch] = useState('');

    function emitVariable() {
        onVariableChange(endPoint);
    };

    /**
     * Catche the value of the TextInput
     * @param {*} value 
     */
    function handleTextChange(value) {
        setTextSearch(value);
        console.log("Fer fetch de: ", value);
    }

    /**
     * Crea la query que sera enviada al component para
     * per utilitzar desde el para possar abans ? si es de inici o & en cas que ja tingui una queri comenÃ§ada
     */
    function saveFilter() {
        let endpointQuery = '';
        let parmaArr = [];

        setModalVisible(false);
        console.log("filter saved");

        //search text
        if (textSearch != '') {
            parmaArr.push('search='+textSearch);
        }

        //query per les tematiques
        if (tematiquesArry().length != 0) {
            let tematicQuery = 'tematiques_nom_in=';
            tematiquesArry().forEach((value, index) => {
                if (index === 0) {
                    tematicQuery+=value;
                }
                else {
                    tematicQuery+=','+value;
                }
            })
            parmaArr.push(tematicQuery);
        }
        
        //query per interval de dates, si les dates son el dia de avui es excepcio i no s'aplica
        if (dateIni.getTime() != dateFi.getTime()) {
            parmaArr.push('dataIni__range=' + dateIni.toISOString() + ',' + dateFi.toISOString());
            parmaArr.push('dataFi__range=' + dateIni.toISOString() + ',' + dateFi.toISOString());
        }
        
        //sha de canviar pero de moment posso limit
        if (slider1 != 0) {
            parmaArr.push('limit='+slider1.toString());
        }

        parmaArr.forEach((value, index) => {
            if (index === 0) {
                endpointQuery+=value;
            }
            else {
                endpointQuery+='&'+value;
            }
        })
        console.log("Final endpoint Query: ", endpointQuery);
        setEndPoint(endpointQuery);
        emitVariable();
    }

    /**
     * Crea i retorna un array de strings de les tematiques seleccionades
     */
    function tematiquesArry() {
        let arr = [];
        if (checkTeatre) {
            arr.push('Teatre');
        }
        if (checkEspectacles) {
            arr.push('Espectacles');
        }
        if (checkMusica) {
            arr.push('Musica');
        }
        if (checkExposicions) {
            arr.push('Exposicions');
        }
        if (checkCinema) {
            arr.push('Cinema');
        }
        if (checkDivulgacio) {
            arr.push('Divulgacio');
        }
        if (checkDansa) {
            arr.push('Dansa');
        }
        if (checkInfantil) {
            arr.push('Infantil');
        }

        return arr;
    }

    const handleDateChangeIni = (event, selectedDate) => {
        const currentDate = selectedDate || dateIni;
        setShowDatePickerIni(false);
        setDateIni(currentDate);
    };

    const handleDateChangeFi = (event, selectedDate) => {
        const currentDate = selectedDate || dateFi;
        setShowDatePickerFi(false);
        setDateFi(currentDate);
    };

    return (
        <View style={styles.contetnView}>
            <Pressable 
                title="filter" 
                onPress={() => setModalVisible(true)}
                style={styles.buttonIconFilter} 
            >
                <FilterLeft name="search" color={'black'} size={30} />
            </Pressable>

            <Modal visible={modalVisible} animationType="slide">
                <View>
                    <Pressable style={styles.closeModal} onPress={() => setModalVisible(false)}>
                        <ArrowDown name="closeModal" color={'black'} size={30} />
                    </Pressable>
                    <View style={styles.contentModal}>
                        <Text style={styles.textModal}>Data inici i final</Text>
                        <View style={styles.contentDataPicker}>
                            <Pressable style={styles.datePicker} onPress={() => setShowDatePickerIni(true)} >
                                <Text> {dateIni.toLocaleDateString()} </Text>
                                <CalendarEvent name="calendarDateIni" color={'black'} size={30} />    
                            </Pressable>
                            {showDatePickerIni && (
                                <DateTimePicker
                                    testID="dateTimePickerIni"
                                    value={dateIni}
                                    mode="date"
                                    is24Hour={true}
                                    display="default"
                                    onChange={handleDateChangeIni}
                                />
                            )}

                            <Pressable style={styles.datePicker} onPress={() => setShowDatePickerFi(true)} >
                                <Text> {dateFi.toLocaleDateString()} </Text>
                                <CalendarEvent name="calendarDateFi" color={'black'} size={30} />    
                            </Pressable>
                            {showDatePickerFi && (
                                <DateTimePicker
                                    testID="dateTimePickerFi"
                                    value={dateFi}
                                    mode="date"
                                    is24Hour={true}
                                    display="default"
                                    onChange={handleDateChangeFi}
                                />
                            )}
                        </View>

                        <Text style={styles.textModal}>Categoria</Text>
                        <CheckBox
                            title="Teatre"
                            checkedColor='green'
                            checked={checkTeatre}
                            onPress={() => setCheckTeatre(!checkTeatre)}
                        />
                        <CheckBox
                            title="Espectacles"
                            checkedColor='green'
                            checked={checkEspectacles}
                            onPress={() => setCheckEspectacles(!checkEspectacles)}
                        />
                        <CheckBox
                            title="Musica"
                            checkedColor='green'
                            checked={checkMusica}
                            onPress={() => setCheckMusica(!checkMusica)}
                        />
                        <CheckBox
                            title="Exposicions"
                            checkedColor='green'
                            checked={checkExposicions}
                            onPress={() => setCheckExposicions(!checkExposicions)}
                        />
                        <CheckBox
                            title="Cinema"
                            checkedColor='green'
                            checked={checkCinema}
                            onPress={() => setCheckCinema(!checkCinema)}
                        />
                        <CheckBox
                            title="Divulgacio"
                            checkedColor='green'
                            checked={checkDivulgacio}
                            onPress={() => setCheckDivulgacio(!checkDivulgacio)}
                        />
                        <CheckBox
                            title="Dansa"
                            checkedColor='green'
                            checked={checkDansa}
                            onPress={() => setCheckDansa(!checkDansa)}
                        />
                        <CheckBox
                            title="Infantil"
                            checkedColor='green'
                            checked={checkInfantil}
                            onPress={() => setCheckInfantil(!checkInfantil)}
                        />

                        {isList ? (<></>) : (
                            <>
                            <Text style={styles.textModal}>Distancia</Text>
                            <Slider
                                value={slider1}
                                onValueChange={setSlider1}
                                minimumValue={0}
                                maximumValue={1000}
                                step={50} 
                            />
                            </>
                        )}
                        
                        
                    </View>
                    <View style={styles.btnFilter}>
                        <Button
                            onPress={saveFilter}
                            title="Filter"
                            color="#2FDD60"
                        />
                    </View>
                    
                </View>
            </Modal>

            <TextInput 
                value={textSearch}
                onBlur={saveFilter}
                onChangeText={handleTextChange}
                placeholder={'Search'}
                placeholderTextColor={'#666'}
            >
            </TextInput>
        </View>
    );
  }

  const styles = StyleSheet.create({
    contetnView: {
        position: 'absolute', 
        top: 10,
        left: 8, 
        width: '90%', 
        borderRadius: 10,
        margin: 10,
        color: '#000',
        borderColor: '#666',
        backgroundColor: '#FFF',
        borderWidth: 1,
        height: 45,
        paddingHorizontal: 10,
        fontSize: 18,
        flexDirection: 'row',
        justifyContent: 'space-between', 
    },
    contentDataPicker: {
        marginLeft: '5%',
        marginRight: '10%',
        marginBottom: '5%',
        marginTop: '5%',
        flexDirection: 'row',
        justifyContent: 'space-between', 
    },
    buttonIconFilter: {
        alignItems: 'center',
        justifyContent: 'center',
        padding: 5,
    },
    btnFilter: {
        width: '50%',
        marginLeft: '25%',
        marginRight: '25%',
        marginTop: '20%',
    },
    closeModal: {
        margin: 15,
    },
    textModal: {
        fontWeight: 'bold',
    },
    datePicker: {
        width: '35%',
        flexDirection: 'row',
        justifyContent: 'space-between', 
    },
    contentModal:{
        margin: 10,
    }
  });