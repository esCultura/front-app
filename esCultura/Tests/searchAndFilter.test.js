import React from 'react';
import SearchFilter from '../src/components/SearchFilter';
import {render, cleanup, fireEvent} from 'react-native-testing-library';

afterEach(cleanup);

describe('<SearchFilter />', () => {

    

    //every it is a test
    it('should match snapshot', () => {
        const onVarChange = jest.fn();
        const comp = render(<SearchFilter onVariableChange={onVarChange} isList={false} />);
        let component = comp.toJSON();
        expect(component).toMatchSnapshot();
    });

    /*
    it('should fire onSubmit events', () => {
        const onVarChange = jest.fn();
        const comp = render(<SearchFilter onVariableChange={onVarChange} isList={false} />);
        const onSubmit = jest.fn();
        const buttonComponent = comp.getByTestId('btnModal');
        
        fireEvent(buttonComponent, 'press');
        
        expect(onSubmit).toHaveBeenCalled();
    });
    */

    /*
    it('should fire onChange events', () => {
        const onChange = jest.fn();
        const { getByTestId } = comp;
      
        const inputComponent = getByTestId('input');
        fireEvent.changeText(inputComponent, 'new text');
      
        expect(onChange).toHaveBeenCalledWith('new text');
      });
    */    

});