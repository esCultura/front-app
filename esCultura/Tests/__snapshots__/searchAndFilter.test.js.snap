// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchFilter /> should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFF",
      "borderColor": "#666",
      "borderRadius": 10,
      "borderWidth": 1,
      "color": "#000",
      "flexDirection": "row",
      "fontSize": 18,
      "height": 45,
      "justifyContent": "space-between",
      "margin": 10,
      "paddingHorizontal": 10,
      "width": 370,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "padding": 5,
      }
    }
    testID="btnModal"
    title="filter"
  >
    <RNSVGSvgView
      bbHeight="16"
      bbWidth="16"
      color="black"
      fill="currentColor"
      focusable={false}
      height={16}
      name="search"
      size={30}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      tintColor="black"
      width={16}
    >
      <RNSVGGroup
        fill={
          {
            "type": 2,
          }
        }
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M2 10.5a.5.5 0 01.5-.5h3a.5.5 0 010 1h-3a.5.5 0 01-.5-.5zm0-3a.5.5 0 01.5-.5h7a.5.5 0 010 1h-7a.5.5 0 01-.5-.5zm0-3a.5.5 0 01.5-.5h11a.5.5 0 010 1h-11a.5.5 0 01-.5-.5z"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="search"
    placeholderTextColor="#666"
    testID="input"
    value=""
  />
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    visible={false}
  />
</View>
`;
